File: backend/src/main/java/me/yeonjae/tonebridge/domain/studycard/StudyCard.java
package me.yeonjae.tonebridge.domain.studycard;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

public record StudyCard(
        UUID id,
        UUID sessionId,
        UUID createdByUserId,
        String phrase,
        String context,
        String nativeAudioUrl,
        String explanation,
        List<String> tags,
        Instant createdAt,
        LearnerAttempt latestAttempt,
        String note
) {
    public boolean hasNativeAudio() {
        return nativeAudioUrl != null && !nativeAudioUrl.isBlank();
    }

    public StudyCard withNativeAudio(String audioUrl) {
        return new StudyCard(id, sessionId, createdByUserId, phrase, context,
                audioUrl, explanation, tags, createdAt, latestAttempt, note);
    }

    public StudyCard withLatestAttempt(LearnerAttempt attempt) {
        return new StudyCard(id, sessionId, createdByUserId, phrase, context,
                nativeAudioUrl, explanation, tags, createdAt, attempt, note);
    }

    public StudyCard withNote(String newNote) {
        return new StudyCard(id, sessionId, createdByUserId, phrase, context,
                nativeAudioUrl, explanation, tags, createdAt, latestAttempt, newNote);
    }
}

File: mobile/lib/features/study_session/domain/model/study_card.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:tonebridge/features/study_session/domain/model/learner_attempt.dart';

part 'study_card.freezed.dart';
part 'study_card.g.dart';

enum CardStatus { noAudio, recorded, corrected }

@freezed
abstract class StudyCard with _$StudyCard {
  const factory StudyCard({
    required String id,
    required String sessionId,
    required String createdByUserId,
    required String phrase,
    String? context,
    String? nativeAudioUrl,
    String? explanation,
    @Default([]) List<String> tags,
    @Default(0) int position,
    required DateTime createdAt,
    LearnerAttempt? latestAttempt,
    String? note,
  }) = _StudyCard;

  const StudyCard._();

  factory StudyCard.fromJson(Map<String, dynamic> json) =>
      _$StudyCardFromJson(json);

  CardStatus get cardStatus {
    if (nativeAudioUrl == null) return CardStatus.noAudio;
    if (latestAttempt?.isCorrected == true) return CardStatus.corrected;
    return CardStatus.recorded;
  }
}
